ipld/retriever.go

```diff
@@ -0,0 +1,178 @@
+	for i := 0; i < size; i++ {
+		roots[i] = plugin.MustCidFromNamespacedSha256(daroots[i])
+	}
+
+	qs := make([]*quadrant, 4) // there are constantly four quadrants
+	for i := range qs {
+		// convert quadrant index to coordinate
+		x, y := i%2, 0
+		if i > 1 {
+			y = 1
+		}
+
+		qs[i] = &quadrant{
+			roots: roots[origSize*y : origSize*(y+1)],
+			x:     x,
+			y:     y,
+		}
+	}
+	// shuffling helps data to be equally distributed around the network
+	rand.Shuffle(len(qs), func(i, j int) { qs[i], qs[j] = qs[j], qs[i] })
+
+	ses := r.newSession(ctx, dah)
+	for _, q := range qs {
+		// TODO: ErrByzantineRow/Col
+		eds, err := ses.retrieve(ctx, q)
+		if err == nil {
+			return eds, nil
+		}
+		// retry quadrants until we have something or error
+	}
+
+	return nil, format.ErrNotFound
+}
+
+type retrieverSession struct {
+	dag   format.NodeGetter
+	adder *NmtNodeAdder
+
+	treeFn rsmt2d.TreeConstructorFn
+	codec  rsmt2d.Codec
+
+	dah    *da.DataAvailabilityHeader
+	square [][]byte
+}
+
+func (r *Retriever) newSession(ctx context.Context, dah *da.DataAvailabilityHeader) *retrieverSession {
+	size := len(dah.RowsRoots)
+	adder := NewNmtNodeAdder(ctx, format.NewBatch(ctx, r.dag, format.MaxSizeBatchOption(batchSize(size))))
+	return &retrieverSession{
+		dag:   merkledag.NewSession(ctx, r.dag),
+		adder: adder,
+		treeFn: func() rsmt2d.Tree {
+			tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(size)/2, nmt.NodeVisitor(adder.Visit))
+			return &tree
+		},
+		codec:  r.codec,
+		dah:    dah,
+		square: make([][]byte, size*size),
+	}
+}
+
+func (rs *retrieverSession) retrieve(ctx context.Context, q *quadrant) (*rsmt2d.ExtendedDataSquare, error) {
+	defer func() {
+		// all shares which were requested or repaired are written to disk with the commit
+		// we store *all*, so they are served to the network, including incorrectly committed data(BEFP case),
+		// so that network can check BEFP
+		if err := rs.adder.Commit(); err != nil {
+			log.Errorw("committing DAG", "err", err)
+		}
+	}()
+	// request and fill quadrant's into
+	// we don't care about the errors here, just need to request as much data as we can to be able to reconstruct below
+	rs.request(ctx, q)
+
+	// try repair
+	err := rsmt2d.RepairExtendedDataSquare(rs.dah.RowsRoots, rs.dah.ColumnRoots, rs.square, rs.codec, rs.treeFn)
+	if err != nil {
+		log.Errorw("not enough shares sampled to recover, retrying...", "err", err)
+		return nil, err
+	}
+
+	return rsmt2d.ImportExtendedDataSquare(rs.square, rs.codec, rs.treeFn)
+}
+
+type quadrant struct {
+	roots []cid.Cid
+	// Coordinates of each quadrant(x;y)
+	// ------  -------
+	// |  Q0 | | Q1  |
+	// |(0:0)| |(1:0)|
+	// ------  -------
+	// |  Q2 | | Q3  |
+	// |(0;1)| |(1;1)|
+	// ------  ------
+	x, y int
+}
+
+func (rs *retrieverSession) request(ctx context.Context, q *quadrant) {
+	ctx, cancel := context.WithTimeout(ctx, RetrieveQuadrantTimeout)
+	defer cancel()
+
+	size := len(q.roots)
+	for i, root := range q.roots {
```

Comment:
May be we can request Leaves for each root in a separate go routine? to increase fetching speed.

Reply:
I’ll add a todo in this pr. Later on, while fixing all the left todos, I all triage the time it may take to implement this, and if it takes more than 30mins, then I will make an issue for this.

Reply:
Fixed

Reply:
The fix was easy, and it matters on block sizes we will work with. The problem with _correctness_ should be resolved by _correct_ testing, which includes checking for the ordering of things in case of concurrency involved.

Reply:
@Wondertan I disagree, especially when the code is as complicated as it is here -- it's already difficult to review and check for correctness as is without adding in the parallelisation. IMO it makes it easier to review when you take the problem one step at a time. I understand it's the "right thing to do" and I do not disagree with the optimisation in general, I just think things like this would belong in a separate, subsequent PR (the way that you've done with others before). I don't think that the argument that go is designed to be used with goroutines justifies including the optimisation in this PR.

Reply:
The tradeoff here is a time we don't have. It is faster to solve multiple related problems simultaneously, rather than spreading those over multiple rounds of PRs and reviews. We are not in the state to follow all the software development best practices (which I agree it is), and we have to sacrifice something to speed up development. Btw, from the git history PoV those are different commits and done iteratively. Advise: always review my PRs commit by commit
> IMO it makes it easier to review when you take the problem one step at a time